clients/ruby/zawgyidetector/lib/zawgyidetector/zawgyi_unicode_markov_model.rb

```diff
@@ -0,0 +1,182 @@
+        marko_chain_index = cp - AFT_CP0 + AFT_OFFSET
+      elsif EXA_CP0 <= cp && cp <= EXA_CP1
+        marko_chain_index = cp - EXA_CP0 + EXA_OFFSET
+      elsif EXB_CP0 <= cp && cp <= EXB_CP1
+        marko_chain_index = cp - EXB_CP0 + EXB_OFFSET
+      elsif SPC_CP0 <= cp && cp <= SPC_CP1
+        marko_chain_index = cp - SPC_CP0 + SPC_OFFSET
+      end
+      marko_chain_index
+    end
+
+    # Runs the given input string on both internal Markov chains and computes the probability of the
+    # string being unicode or zawgyi.
+    # @param1 input The string to evaluate.
+    # @param2 verbose Whether to print the log probabilities for debugging.
+    # @return The probability that the string is Zawgyi given that it is either Unicode or Zawgyi, or
+    # -Infinity if there are no Myanmar range code points in the string.
+    def predict(input, verbose=false)
+      if verbose
+        puts "Running detector on string: #{input}"
+      end
+  
+      # Start at the base state
+      prev_cp = 0
+      prev_state = 0
+  
+      total_delta = 0.0
+      seen_transition = false
+  
+      offset = 0
+      while offset <= input.length
+        if offset == input.length
+          cp = 0
+          curr_state = 0
+        else
+          cp = input.codepoints[offset]
+          curr_state = self.class.get_index_for_code_point(cp)
+        end
+  
+        # Ignore 0-to-0 transitions
+        if prev_state != 0 || curr_state != 0
+          # Gets the difference in log probabilities between chain A and chain B.
+          # First param: The index of the source node to transition from.
+          # Second param: The index of the destination node to transition to.
+          delta = @classifier.log_probability_differences[prev_state][curr_state].to_f
+  
+          if verbose
+            puts "#{prev_cp} -> #{cp}: delta=#{delta}"
+            puts "ABS: #{delta.abs}"
+            delta_index = 1
+            while delta_index < delta.abs
+              print "!"
+              delta_index +=1
+            end
+            puts ""
+          end
+  
+          total_delta += delta
+          seen_transition = true
+        end
+  
+        offset += char_count(cp)
+        prev_cp = cp
+        prev_state = curr_state
+      end
+  
+      if verbose
+        puts "Final: delta=#{total_delta}"
+      end
+  
+      # Special case: if there is no signal, return -Infinity,
+      # which will get interpreted by users as strong Unicode.
+      # This happens when the input string contains no Myanmar-range code points.
+      unless seen_transition
+        return -1.0/0.0
+      end
+      1.0 / (1.0 + Math.exp(total_delta))
+    end
+
+    # Determine the number of char values needed to represent the specified character (Unicode code point)
+    # if the code point is equal to or greater than 0x10000, then the method returns 2.
+    # otherwise, the method returns 1.
+    def char_count(code_point)
+      code_point >= 10000 ? 2 : 1
```

Comment:
This should be `0x10000`, not `10000`

Reply:
d'oh! I will replace **10000** with **0x10000**.